### PR TITLE
[Git] - Adding a way to perform the git operations on a clone

### PR DESCRIPTION
**Pull request by acabarbaye**
This PR introduces a new utility `GitTempClone` which wraps the `GitWrapper` in order to perform git actions on a "clone" repository created in the temporary folder so that the initial git repository is not altered.
Git facilities such as `assert_news` and  `tag_and_release` were changed to use this facility

### CODE REVIEW

File: mbed_tools_ci/utils/filesystem_helpers.py [hunk 1]
```diff
@@ -77,3 +80,52 @@ def find_file_in_tree(file_name: str,
             f'File [{file_name}] not found anywhere in directories {"above" if top else "under"} {starting_point}'
         )
     return file_path
+
+
+class TemporaryDirectory:
```
Comment (rwalton-arm) [inline mbed_tools_ci/utils/filesystem_helpers.py:86]: Have you seen https://docs.python.org/3/library/tempfile.html#tempfile.TemporaryDirectory ? It might cover our usecase here.

Comment (rwalton-arm) [reply mbed_tools_ci/utils/filesystem_helpers.py:86]: After our conversation it seems `TemporaryDirectory` doesn't work on windows with certain files, it fails on cleanup. Which is annoying. There's an issue on the python bug-tracker from 2016: https://bugs.python.org/issue26660

Comment (acabarbaye) [reply mbed_tools_ci/utils/filesystem_helpers.py:86]: I referenced this issue in the doc string so that we remember the rationale

File: mbed_tools_ci/utils/git_helpers.py [hunk 2]
```diff
@@ -36,13 +48,56 @@ def _git_url_ssh_to_https(self, url: str) -> str:
...
+        return self._root_path
+
     def configure_author(self) -> None:
         """Sets the author."""
         self.repo.config_writer().set_value("user", "name",
                                             self.author.name).release()
         self.repo.config_writer().set_value("user", "email",
                                             self.author.email).release()
 
+    def checkout_branch_from_name(self, branch_name: str) -> Any:
```
Comment (muchzill4) [inline mbed_tools_ci/utils/git_helpers.py]: `checkout_branch` would be enough

File: mbed_tools_ci/utils/git_helpers.py [hunk 3]
```diff
@@ -159,10 +223,25 @@ def get_current_branch(self) -> Any:
...
+            current_branch = self.repo.git.branch(show_current=True)
+            current_branch = current_branch if isinstance(current_branch,
+                                                          str) else current_branch.decode(
+                'utf-8')
+            return self.get_branch(current_branch)
+        return None
+
+    def _get_branch_from_abbreviation(self, abbreviation: str) -> Any:
+        current_branch = self.repo.git.rev_parse('--abbrev-ref', abbreviation)
+        current_branch = current_branch if isinstance(current_branch,
```
Comment (muchzill4) [inline mbed_tools_ci/utils/git_helpers.py:241]: Is `rev_parse` really returning inconsistent types here? (bytes and str)

File: mbed_tools_ci/utils/git_helpers.py [hunk 4]
```diff
@@ -159,10 +223,25 @@ def get_current_branch(self) -> Any:
...
             )
-        return self.get_branch(
-            self.repo.git.rev_parse('--abbrev-ref', 'HEAD').decode(
-                'utf-8').strip()
-        )
+        current_branch = self._get_branch_from_advanced_feature()
+        return current_branch if current_branch else self._get_branch_from_abbreviation(
+            'HEAD')
+
+    def _get_branch_from_advanced_feature(self) -> Any:
```
Comment (muchzill4) [inline mbed_tools_ci/utils/git_helpers.py:230]: Is this necessary? Can we not rely on one way of getting branches?

File: mbed_tools_ci/utils/git_helpers.py [hunk 5]
```diff
@@ -159,10 +223,25 @@ def get_current_branch(self) -> Any:
             logger.warning(
                 f'Could not determine the branch name using GitPython: {e}'
             )
-        return self.get_branch(
-            self.repo.git.rev_parse('--abbrev-ref', 'HEAD').decode(
-                'utf-8').strip()
-        )
+        current_branch = self._get_branch_from_advanced_feature()
+        return current_branch if current_branch else self._get_branch_from_abbreviation(
```
Comment (muchzill4) [inline mbed_tools_ci/utils/git_helpers.py]: Nit: I find those inline `if else` unreadable most of the time, especially in this case, where the argument to the function is wrapped on the following line.

```
current_branch = self._get_it_the_first_way()
if not current_branch:
    current_branch = self._get_it_some_other_way()
return current_branch
```

File: tests/git_helper/test_git_helpers.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+        self.assertIsNotNone(git.get_remote_url())
+
+
+class TestGitTempClone(TestCase):
+
+    def test_git_clone(self):
+        """Ensures a fully fledged clone is created."""
+        with GitTempClone(current_branch_name='master') as clone:
+            self.assertIsNotNone(clone)
+            self.assertTrue(isinstance(clone, GitWrapper))
```
Comment (muchzill4) [inline tests/git_helper/test_git_helpers.py:32]: This and the test above are redundant.

File: tests/git_helper/test_git_helpers.py [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+            self.assertEqual('master', str(clone.get_current_branch()))
+            self.assertFalse(clone.is_current_branch_feature())
```
Comment (muchzill4) [inline tests/git_helper/test_git_helpers.py]: How is this relevant for this test?

File: mbed_tools_ci/assert_news.py [hunk 8]
```diff
@@ -118,20 +120,20 @@ def main() -> None:
...
-        try:
-            validate_news_files(
-                git=git,
-                news_dir=news_dir,
-                root_dir=root_dir,
-            )
-        except Exception as e:
-            log_exception(logger, e)
-            sys.exit(1)
+    with GitTempClone(args.current_branch)as git:
```
Comment (muchzill4) [inline mbed_tools_ci/assert_news.py]:
```suggestion
    with GitTempClone(args.current_branch) as git:
```

File: mbed_tools_ci/assert_news.py [hunk 9]
```diff
@@ -108,6 +108,8 @@ def main() -> None:
     parser = argparse.ArgumentParser(
         description="Check correctly formatted news files exist on feature branch."
     )
+    parser.add_argument('-b', '--current-branch',
```
Comment (muchzill4) [inline mbed_tools_ci/assert_news.py:113]: It seems odd that we have to pass this as argument. Shouldn't git wrapper be able to figure this out on its own? You've added `get_branch_from_*` methods. 🤔

File: mbed_tools_ci/utils/git_helpers.py [hunk 10]
```diff
@@ -462,3 +554,76 @@ def is_current_branch_feature(self) -> bool:
...
+            else:
+                logger.info(f"Removing {destination}")
+                if destination.exists():
+                    destination.unlink()
+
+
+class GitTempClone:
+    """Context manager providing a temporary cloned repository."""
+
+    def __init__(self, current_branch_name: Optional[str] = None,
```
Comment (muchzill4) [inline mbed_tools_ci/utils/git_helpers.py]: 1. This is never called with `repository_to_clone` other than in test. 
Also, I would avoid the pattern of magic defaults (cause `Optional[]` stinks!) and wrap those use cases in specific functions or classes. Otherwise, I think this is walking on the edge of single responsibility principle.
My suggestion:
```
class GitTempClone:
    def __init__(self, repository_to_clone: GitWrapper):  # this always wants repository_to_clone
        self._clone = repository_to_clone

class ProjectRootGitTempClone(GitTempClone):
    def __init__(self):
        git = GitWrapper(configuration.get_value(ConfigurationVariable.PROJECT_ROOT))
        super().__init__(repository_to_clone=git) 
```

2. `temporary_directory` is never passed in, why have it as argument?

3. I think the branching should not happen here. Same with `apply_uncomitted_changes` (what does it even do?). Just do a clone.

4. Finally, I think you're better off using more succinct:
```
@contextlib.contextmanager
def git_temp_clone(repository: GitWrapper):
    with TemporaryDirectory() as temp_dir:
        clone = repository.cone(temp_dir.path)
        yield clone

@contextlib.contextmanager
def git_temp_clone_of_project_root():
    git = GitWrapper(configuration.get_value(ConfigurationVariable.PROJECT_ROOT))
    with git_temp_clone(git) as clone:
        yield clone
```

File: mbed_tools_ci/utils/filesystem_helpers.py [hunk 11]
```diff
@@ -77,3 +81,56 @@ def find_file_in_tree(file_name: str,
...
+
+    This is the same as tempfile.TemporaryDirectory in most cases.
+    However, on CI the temporary folder may not be accessible.
+    Moreover, on Windows TemporaryDirectory will fail cleaning up
+    (https://bugs.python.org/issue22107) and hence, a workaround had to be
+    implemented.
+
+    """
+
+    def __init__(self, path: Optional[str] = None) -> None:
```
Comment (muchzill4) [inline mbed_tools_ci/utils/filesystem_helpers.py]: I don't think this should be concerned with a path. I'd go with "Hey `TempDirectory` just give me a valid temp directory, I don't care where it is".

File: mbed_tools_ci/utils/filesystem_helpers.py [hunk 12]
```diff
@@ -77,3 +81,56 @@ def find_file_in_tree(file_name: str,
...
+        """Context manager exit point."""
+        self.cleanup()
+
+    def cleanup(self) -> None:
+        """Deletes the temporary directory."""
+        if self._tmp_dir_context_manager and not (
+                platform.system() == 'Windows'):
+            self._tmp_dir_context_manager.cleanup()
+        else:
+            shutil.rmtree(self._tmp_dir_path, ignore_errors=True)
```
Comment (muchzill4) [inline mbed_tools_ci/utils/filesystem_helpers.py:124]: Could we not always do this instead of doing `Windows` check?

File: mbed_tools_ci/utils/filesystem_helpers.py [hunk 13]
```diff
@@ -77,3 +81,56 @@ def find_file_in_tree(file_name: str,
...
+        """String representation."""
+        return str(self._tmp_dir_path)
+
+    def __enter__(self) -> Path:
+        """Context manager entry point."""
+        return self.path
+
+    def __exit__(self, exc: Any, value: Any, tb: Any) -> None:
+        """Context manager exit point."""
+        self.cleanup()
```
Comment (muchzill4) [inline mbed_tools_ci/utils/filesystem_helpers.py:117]: Why not put `cleanup` body in here?

File: mbed_tools_ci/utils/git_helpers.py [hunk 14]
```diff
@@ -462,3 +554,76 @@ def is_current_branch_feature(self) -> bool:
...
+        )
+        if not status:
+            return []
+
+        return [
+            Path(self.root).joinpath(
+                line.strip().split(' ')[-1]) for line in status.splitlines()
+        ]
+
+    def apply_uncommitted_changes(self, other_repo: Union[Path, 'GitWrapper']) -> None:
```
Comment (muchzill4) [inline mbed_tools_ci/utils/git_helpers.py]: From what I've seen this is only ever called with `GitWrapper`, can we get rid of `Path` type?

**Review by muchzill4 — CHANGES_REQUESTED** [review]
I really like that we operate on the clone now. It is much better in comparison to previous version where we'd blat our local repositories if this code failed for some reason.

Dropping some questions and feedback.

File: mbed_tools_ci/utils/git_helpers.py (hunk 4 above)
Comment (acabarbaye) [reply mbed_tools_ci/utils/git_helpers.py:230]: there is not a relying way:
it depends on the git version you are using and on the system you are running the thing on:
these are two things we have no control on

File: mbed_tools_ci/utils/git_helpers.py (hunk 5 above)
Comment (acabarbaye) [reply mbed_tools_ci/utils/git_helpers.py]: We'll change for ease of read

File: mbed_tools_ci/utils/git_helpers.py (hunk 3 above)
Comment (acabarbaye) [reply mbed_tools_ci/utils/git_helpers.py:241]: yes unfortunately.
At least from what I experienced

File: mbed_tools_ci/utils/git_helpers.py (hunk 10 above)
Comment (acabarbaye) [reply mbed_tools_ci/utils/git_helpers.py]: 1 and 2 agreed
3 the thing only works because of apply_uncomitted_changes. changes happen in the initial repo and the clone is created to commit them back to the remote repo (this is what tag and release does) so the clone needs to get those changes
4 I don't like this style and prefer the OOP style which I actually find clearer

File: mbed_tools_ci/utils/filesystem_helpers.py (hunk 11 above)
Comment (acabarbaye) [reply mbed_tools_ci/utils/filesystem_helpers.py]: I removed it but it may need to be reintroduced in the future if we have limitation in some CI systems. This would be a conversation for later

File: mbed_tools_ci/utils/filesystem_helpers.py (hunk 13 above)
Comment (acabarbaye) [reply mbed_tools_ci/utils/filesystem_helpers.py:117]: just did it

File: mbed_tools_ci/utils/git_helpers.py (hunk 10 above)
Comment (muchzill4) [reply mbed_tools_ci/utils/git_helpers.py]: Re 3.
If that's the case we should clone first and then do the changes in the cloned repo only. At the moment _random_ external state is leaked into this class. Ideally we should be able to use it on its own w/o all the burden of how our CI works. :)

I do understand if that's not possible.

File: mbed_tools_ci/utils/git_helpers.py (hunk 14 above)
Comment (acabarbaye) [reply mbed_tools_ci/utils/git_helpers.py]: done

File: mbed_tools_ci/utils/git_helpers.py (hunk 10 above)
Comment (acabarbaye) [reply mbed_tools_ci/utils/git_helpers.py]: the utility is to mess up with a git repo which is not the primary repository. other scripts or CI steps which have no knowledge of git or source control should not be aware of this mechanism and of the existence of a clone. this is why this is done this way and what I absolutely don't want is to put the burden of git management to other scripts: it is not their concerns.

File: mbed_tools_ci/utils/filesystem_helpers.py (hunk 13 above)
Comment (acabarbaye) [reply mbed_tools_ci/utils/filesystem_helpers.py:117]: Had to roll back this change as cleanup method is called by the GitTempClone class and this method is clearer to call thant .__exit__(.....)

File: mbed_tools_ci/assert_news.py (hunk 9 above)
Comment (acabarbaye) [reply mbed_tools_ci/assert_news.py:113]: It is unfortunably not possible. On azure pipelines, the information about the current branch is solely contained in Environment variables:  depending if the trigger is a branch or a pull request, either $(Build.SourceBranchName) or $(System.PullRequest.SourceBranch)

Same for CircleCI, $CIRCLE_BRANCH

File: mbed_tools_ci/utils/git_helpers.py (hunk 10 above)
Comment (muchzill4) [reply mbed_tools_ci/utils/git_helpers.py]: I'm not saying we should put this in our script files. I'm saying this does a bit much by assuming the state of the "cloned" repo. The fact that there's this global state (git changes) we have to carry around is not great, as you have no control over it and need to replicate it. In the ideal world you shouldn't.

I.e., the release process should be something like:
```
with GitTempClone(git=GitWrapper(PROJECT_ROOT)) as git_clone:
    # git and git_clone should be CLEAN at this point, why isn't it and we need to `apply_uncommited_changes`?
    update_documentation(git_clone)
    git.add('_version', 'changelog')
    git.commit('release!')
    git.tag()
    git.push()
```

File: mbed_tools_ci/assert_news.py [hunk 15]
```diff
@@ -58,7 +58,7 @@ def validate_news_file(absolute_path: str) -> None:
 
 
 def find_news_files(
-    git: GitWrapper, root_dir: str, news_dir: str
+        git: GitWrapper, root_dir: str, news_dir: str
```
Comment (muchzill4) [inline mbed_tools_ci/assert_news.py:61]: Black will tell you this is bad. :trollface:

File: mbed_tools_ci/assert_news.py [hunk 16]
```diff
@@ -73,12 +73,14 @@ def find_news_files(
...
-    return [str(pathlib.Path(root_dir, file_path)) for file_path in added_news_files]
+    added_news_files = [file_path for file_path in files_changed if
+                        file_path.startswith(news_dir)]
+    return [str(pathlib.Path(root_dir, file_path)) for file_path in
+            added_news_files]
 
 
 def validate_news_files(
-    git: GitWrapper, root_dir: str, news_dir: str
+        git: GitWrapper, root_dir: str, news_dir: str
```
Comment (muchzill4) [inline mbed_tools_ci/assert_news.py:83]: Black will tell you this is bad. :trollface:

File: mbed_tools_ci/assert_news.py [hunk 17]
```diff
@@ -118,20 +122,23 @@ def main() -> None:
...
-        try:
-            validate_news_files(
-                git=git,
-                news_dir=news_dir,
-                root_dir=root_dir,
-            )
-        except Exception as e:
-            log_exception(logger, e)
-            sys.exit(1)
+    with ProjectTempClone(current_branch_name=args.current_branch)as git:
```
Comment (muchzill4) [inline mbed_tools_ci/assert_news.py]: I think in `ProjectTempClone`s world it's `desired_branch_name` and not `current_branch_name`. Thoughts?

File: mbed_tools_ci/assert_news.py [hunk 18]
```diff
@@ -118,20 +122,23 @@ def main() -> None:
...
+        if git.is_current_branch_feature():
+            root_dir = configuration.get_value(
```
Comment (muchzill4) [inline mbed_tools_ci/assert_news.py:127]: Nit: This wrapping is nuts, brackets should be indented on the same level. There's no need for a line break here and in the lines below.

File: mbed_tools_ci/tag_and_release.py [hunk 19]
```diff
@@ -76,27 +82,29 @@ def _update_documentation(git: GitWrapper) -> None:
...
+            git.add(
+                configuration.get_value(
+                    ConfigurationVariable.VERSION_FILE_PATH))
+            git.add(
+                configuration.get_value(
+                    ConfigurationVariable.CHANGELOG_FILE_PATH))
+            git.add(configuration.get_value(ConfigurationVariable.NEWS_DIR))
+            time_str = datetime.datetime.utcnow().strftime(
+                "%Y-%m-%d %H:%M")
+            commit_message = f':rocket: :newspaper: releasing version {version} @ {time_str}' if is_new_version else f':gear: Automatic changes'
```
Comment (muchzill4) [inline mbed_tools_ci/tag_and_release.py]: Just realised, we should probably use legit emoji images here instead of `:foo:`.

File: mbed_tools_ci/utils/filesystem_helpers.py [hunk 20]
```diff
@@ -77,3 +81,50 @@ def find_file_in_tree(file_name: str,
...
+
+    def __init__(self) -> None:
+        """Constructor."""
+        self._tmp_dir_context_manager = tempfile.TemporaryDirectory()
+        _path = self._tmp_dir_context_manager.name
+        if not _path:
+            raise ValueError("Undefined temporary directory.")
+        self._tmp_dir_path = Path(_path).resolve()
+
+        if not self._tmp_dir_path.exists():
```
Comment (muchzill4) [inline mbed_tools_ci/utils/filesystem_helpers.py]: When would that ever be the case - worth commenting? 
If ever, why not always `self._tmp_dir.path.mkdir(exists_ok=True)`?

File: mbed_tools_ci/utils/filesystem_helpers.py [hunk 21]
```diff
@@ -77,3 +81,50 @@ def find_file_in_tree(file_name: str,
...
+    (https://bugs.python.org/issue22107) and hence, a workaround had to be
+    implemented.
+
+    """
+
+    def __init__(self) -> None:
+        """Constructor."""
+        self._tmp_dir_context_manager = tempfile.TemporaryDirectory()
+        _path = self._tmp_dir_context_manager.name
+        if not _path:
```
Comment (muchzill4) [inline mbed_tools_ci/utils/filesystem_helpers.py]: When would that be the case? Maybe comment? According to `TemporaryDirectory()` docs it should always give you a `.name` back.

File: mbed_tools_ci/utils/git_helpers.py [hunk 22]
```diff
@@ -462,3 +556,87 @@ def is_current_branch_feature(self) -> bool:
...
+    def uncommitted_changes(self) -> List[Path]:
+        """Gets list of uncommitted files.
+
+        Returns:
+             list of uncommitted
+        """
+        status = self.repo.git.status(
+            porcelain=True,
+            untracked_files=True
+
```
Comment (muchzill4) [inline mbed_tools_ci/utils/git_helpers.py]:
```suggestion
```

File: mbed_tools_ci/utils/git_helpers.py [hunk 23]
```diff
@@ -436,6 +521,15 @@ def create_branch(self, branch_name: str) -> Any:
         """
         return self.repo.create_head(branch_name)
 
+    def git_version(self) -> str:
+        """Gets git version.
+
+        Returns:
+            the version of git in use.
+        """
+        return '.'.join(
```
Comment (muchzill4) [inline mbed_tools_ci/utils/git_helpers.py:528]: Nit: unnecessary line wrapping

File: mbed_tools_ci/utils/git_helpers.py [hunk 24]
```diff
@@ -2,21 +2,33 @@
...
-            configuration.get_value(ConfigurationVariable.PROJECT_ROOT))
+    def __init__(self, path: Optional[str] = None,
+                 repo: Optional[Repo] = None) -> None:
+        """Creates an instance of GitWrapper.
+
+        Args:
+            path: path to repository. If None, project root is considered.
+            repo: GitPython repository. if None, the repository is populated.
+        """
+        self._root_path = Path(path if path else configuration.get_value(
```
Comment (muchzill4) [inline mbed_tools_ci/utils/git_helpers.py]: I think this should take similar treatment as `GitTempClone` and `ProjectTempClone` had. I.e.: separate construction of a known object to a separate class. Less braching in `__init__` is always a win.

**Review by muchzill4 — CHANGES_REQUESTED** [review]
Thanks for addressing my suggestions. I'm leaving a bit more feedback. 👍

File: mbed_tools_ci/assert_news.py (hunk 17 above)
Comment (acabarbaye) [reply mbed_tools_ci/assert_news.py]: agreed

File: mbed_tools_ci/tag_and_release.py (hunk 19 above)
Comment (acabarbaye) [reply mbed_tools_ci/tag_and_release.py]: changing so that git logs are no longer offending ;)

File: mbed_tools_ci/tag_and_release.py [hunk 25]
```diff
@@ -97,7 +97,7 @@ def _update_repository(mode: CommitType, is_new_version: bool,
             git.add(configuration.get_value(ConfigurationVariable.NEWS_DIR))
             time_str = datetime.datetime.utcnow().strftime(
                 "%Y-%m-%d %H:%M")
-            commit_message = f':rocket: :newspaper: releasing version {version} @ {time_str}' if is_new_version else f':gear: Automatic changes'
+            commit_message = f'📰 releasing version {version}🚀 @ {time_str}' if is_new_version else f'📰 Automatic changes ⚙'
```
Comment (muchzill4) [inline mbed_tools_ci/tag_and_release.py]:
```suggestion
            commit_message = f'📰 releasing version {version} 🚀 @ {time_str}' if is_new_version else f'📰 Automatic changes ⚙'
```

**Review by muchzill4 — APPROVED** [review]
Nice work. 👍
